### PR TITLE
core: use LRU cache in txNoncer to avoid memory leaks

### PR DESCRIPTION
**Pull request by dbadoy**
Hi,
This is the PR that use LRU cache instead of map in txNoncer.
The original codes could cause a memory leak in txNoncer.

Send rpc request `eth_getTransactionCount` with `"pending"`, It'll call txNoncer.get().
```go
// txNoncer.get
if _, ok := txn.nonces[addr]; !ok {
	txn.nonces[addr] = txn.fallback.GetNonce(addr)
}
```
If there is no address in `txNoncer`, get the nonce from `stateDB`.
`stateDB.GetNonce` returns the nonce value of the target address, or uint64(0) if it does not exist.
It's means that the address:nonce pair is always stored in txNonce map even not exist address in DB.

```go
// The original codes
func (txn *txNoncer) get(addr common.Address) uint64 {
	// We use mutex for get operation is the underlying
	// state will mutate db even for read access.
	txn.lock.Lock()
	defer txn.lock.Unlock()

	if _, ok := txn.nonces[addr]; !ok {
		txn.nonces[addr] = txn.fallback.GetNonce(addr)
	}
        // Add for test
	fmt.Printf("in txNoncer.get, map size: %d", len(txn.nonces))
	return txn.nonces[addr]
}
```
```typescript
// The Scripts causing memory leaks
while(1) {
    let id = crypto.randomBytes(32).toString('hex');
    let privateKey = "0x"+id;
    let wallet = new ethers.Wallet(privateKey);
    console.log("Address: " + wallet.address)
    await provider.getTransactionCount(wallet.address, "pending");
 }
```
```
.
.
.
in txNoncer.get, map size: 32449
in txNoncer.get, map size: 32450
in txNoncer.get, map size: 32451
in txNoncer.get, map size: 32452
in txNoncer.get, map size: 32453
in txNoncer.get, map size: 32454
in txNoncer.get, map size: 32455
in txNoncer.get, map size: 32456
in txNoncer.get, map size: 32457
in txNoncer.get, map size: 32458
in txNoncer.get, map size: 32459
in txNoncer.get, map size: 32460
in txNoncer.get, map size: 32461
in txNoncer.get, map size: 32462
in txNoncer.get, map size: 32463
in txNoncer.get, map size: 32464
in txNoncer.get, map size: 32465
in txNoncer.get, map size: 32466
in txNoncer.get, map size: 32467
in txNoncer.get, map size: 32468
.
.
.
```
If I keep running this scripts, memory leak will occurs.

In my test case,
```go
func TestTxNoncerMap(t *testing.T) {
	statedb, _ := state.New(common.Hash{}, state.NewDatabase(rawdb.NewMemoryDatabase()), nil)
	txNoncer := newTxNoncerMap(statedb)

	var m runtime.MemStats
	runtime.GC()
	for i := 0; i < 10000000; i++ {
		var b [20]byte
		rand.Read(b[:])
		addr := common.Address(b)
		txNoncer.set(addr, uint64(0))
	}
	runtime.ReadMemStats(&m)
	fmt.Printf("Object memory: %.3f MB current\n", float64(m.Alloc)/1024/1024)
	fmt.Printf("System memory: %.3f MB current\n", float64(m.Sys)/1024/1024)
	fmt.Printf("Allocations:   %.3f million\n", float64(m.Mallocs)/1000000)
}
```
```
Object memory: 807.569 MB current
System memory: 1074.236 MB current
Allocations:   0.333 million
```
I think it's a number worth considering.

<br>

(+)
```go
const nonceCacheLimit = 1024 * 50
```
This cache size places a memory limit of 10MB to 20MB in txNoncer

### CODE REVIEW

**Comment by dbadoy** [comment]
Performance Comparison

```
Store
------------------------------------------------------------------------------------------------
Origin 
BenchmarkMapStore-10             1000000               245.1 ns/op           144 B/op          0 allocs/op
LRU 
BenchmarkLRUStore-10             1000000               312.8 ns/op           115 B/op          3 allocs/op

Load
------------------------------------------------------------------------------------------------
Origin
BenchmarkMapLoad-10      1000000                24.61 ns/op            0 B/op          0 allocs/op
LRU
BenchmarkLRULoad-10      1000000                25.24 ns/op            0 B/op          0 allocs/op
```

File: core/tx_noncer.go [hunk 1]
```diff
@@ -45,43 +49,40 @@ func newTxNoncer(statedb *state.StateDB) *txNoncer {
...
 	// state will mutate db even for read access.
-	txn.lock.Lock()
-	defer txn.lock.Unlock()
-
-	if _, ok := txn.nonces[addr]; !ok {
-		txn.nonces[addr] = txn.fallback.GetNonce(addr)
+	if _, ok := txn.nonces.Get(addr); !ok {
+		txn.lock.Lock()
+		txn.nonces.Add(addr, txn.fallback.GetNonce(addr))
+		txn.lock.Unlock()
```
Comment (holiman) [inline core/tx_noncer.go:55]: >  fixed-size thread safe LRU cache

```suggestion
		txn.nonces.Add(addr, txn.fallback.GetNonce(addr))
```
This should suffice, no?

Comment (holiman) [reply core/tx_noncer.go:55]: Or maybe simpler:
```golang
if nonce, ok := txn.nonces.Get(addr); ok {
     nonce.(uint64)
}
nonce := txn.fallback.GetNonce(addr)
txn.nonces.Add(addr,nonce)
return nonce
```

File: core/tx_noncer.go [hunk 2]
```diff
@@ -45,43 +49,40 @@ func newTxNoncer(statedb *state.StateDB) *txNoncer {
...
+// setIfLower updates a new virtual nonce into the LRU cache if the
 // the new one is lower.
 func (txn *txNoncer) setIfLower(addr common.Address, nonce uint64) {
-	txn.lock.Lock()
-	defer txn.lock.Unlock()
-
-	if _, ok := txn.nonces[addr]; !ok {
-		txn.nonces[addr] = txn.fallback.GetNonce(addr)
+	if _, ok := txn.nonces.Get(addr); !ok {
+		txn.lock.Lock()
```
Comment (holiman) [inline core/tx_noncer.go:71]: Again, no need for locking right?

File: core/tx_noncer_test.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+		addr := common.Address(b)
+		txNoncer.nonces.Add(addr, uint64(1))
+	}
+	runtime.ReadMemStats(&m)
+	fmt.Printf("Object memory: %.3f MB current\n", float64(m.Alloc)/1024/1024)
+	fmt.Printf("System memory: %.3f MB current\n", float64(m.Sys)/1024/1024)
+	fmt.Printf("Allocations:   %.3f million\n", float64(m.Mallocs)/1000000)
+}
+
+func BenchmarkMapStore(b *testing.B) {
```
Comment (holiman) [inline core/tx_noncer_test.go:96]: I don't think we need these benchmarks, I'd personally skip them.

**Review by holiman — COMMENTED** [review]
In general, I agree that it's not "right" that we keep piling stuff indefinitely into a map. However, I'm also not sure about the expected semantics of the noncer -- how callers will behave when it starts forgetting nonces. 
I suppose it should be fine, but need to double-check.

File: core/tx_noncer.go (hunk 1 above)
Comment (dbadoy) [reply core/tx_noncer.go]: I think method that return once would be good.

**Comment by dbadoy** [comment]
@holiman Thanks the review! I applied what you said.
> how callers will behave when it starts forgetting nonces. I suppose it should be fine, but need to double-check.

My concern is...
If starts forgetting nonces, It will use less memory but could be increase the number of StateDB.GetNonce calls.
imho, it's the cache limit that counts.

File: core/tx_pool.go [hunk 4]
```diff
@@ -463,9 +463,6 @@ func (pool *TxPool) SetGasPrice(price *big.Int) {
 // Nonce returns the next nonce of an account, with all transactions executable
 // by the pool already applied on top.
 func (pool *TxPool) Nonce(addr common.Address) uint64 {
```
[the remark this answers is not in the document]
Comment (dbadoy) [reply core/tx_pool.go:465]: ah, I totally get it.

**Comment by dbadoy** [comment]
> If we switch a map for a lossy cache, then we break the semantics, because it will "best effort" and this method may suddenly return the next nonce without pool txs applied on top.

I think LRU cache is not suitable because as you said, the semantics changed...
Back to the purpose of this PR, I tried resolve the memory leak by call `eth_getTransactionCount with "pending"` with  not exist address in DB (not exist means, not used address).

It seems to be possible to solve it to some extent as below,
```go
func (txn *txNoncer) get(addr common.Address) uint64 {
	// We use mutex for get operation is the underlying
	// state will mutate db even for read access.
	txn.lock.Lock()
	defer txn.lock.Unlock()

	if _, ok := txn.nonces[addr]; !ok {
		nonce := txn.fallback.GetNonce(addr)
                //  ## Save only addresses that exist in DB
		if nonce != 0 {
			txn.nonces[addr] = txn.fallback.GetNonce(addr)
		}
	}
	return txn.nonces[addr]
}
```
what do you think about this?

**Comment by dbadoy** [comment]
Thanks for the review. I applied that!

**Review by holiman — APPROVED** [review]
LGTM

**Pull request merged**